From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Richard Townsend <richard.townsend@arm.com>
Date: Wed, 24 Apr 2019 13:57:36 +0100
Subject: Workaround an undefined symbol error

Previously, builds configured with dcheck_always_on=true would error
with messages like this in the log:
 lld-link: error: undefined symbol: public: bool __cdecl
  v8::internal::CPURegister::IsZero(void) const

By moving some functions out of the the arm64-assembler header file,
this error no longer seems to happen.

diff --git a/src/codegen/arm64/assembler-arm64.cc b/src/codegen/arm64/assembler-arm64.cc
index 90f7c3797acf659a65b2eeea933a3362b339359c..bc9ba459550f99f8e3e07b146ce931a6d10c6c1e 100644
--- a/src/codegen/arm64/assembler-arm64.cc
+++ b/src/codegen/arm64/assembler-arm64.cc
@@ -3609,6 +3609,22 @@ void Assembler::MoveWide(const Register& rd, uint64_t imm, int shift,
        ImmMoveWide(static_cast<int>(imm)) | ShiftMoveWide(shift));
 }
 
+Instr Assembler::RmNot31(CPURegister rm) {
+  DCHECK_NE(rm.code(), kSPRegInternalCode);
+  DCHECK(!rm.IsZero());
+  return Rm(rm);
+}
+
+Instr Assembler::RdSP(Register rd) {
+  DCHECK(!rd.IsZero());
+  return (rd.code() & kRegCodeMask) << Rd_offset;
+}
+
+Instr Assembler::RnSP(Register rn) {
+  DCHECK(!rn.IsZero());
+  return (rn.code() & kRegCodeMask) << Rn_offset;
+}
+
 void Assembler::AddSub(const Register& rd, const Register& rn,
                        const Operand& operand, FlagsUpdate S, AddSubOp op) {
   DCHECK_EQ(rd.SizeInBits(), rn.SizeInBits());
diff --git a/src/codegen/arm64/assembler-arm64.h b/src/codegen/arm64/assembler-arm64.h
index f0d6f583a30e37935ad2f68cf44d0b4136e27e2d..dd15f23242a9b5d8a5a1e95875902dc9ca4a7618 100644
--- a/src/codegen/arm64/assembler-arm64.h
+++ b/src/codegen/arm64/assembler-arm64.h
@@ -2119,11 +2119,7 @@ class V8_EXPORT_PRIVATE Assembler : public AssemblerBase {
     return rm.code() << Rm_offset;
   }
 
-  static Instr RmNot31(CPURegister rm) {
-    DCHECK_NE(rm.code(), kSPRegInternalCode);
-    DCHECK(!rm.IsZero());
-    return Rm(rm);
-  }
+  static Instr RmNot31(CPURegister rm);
 
   static Instr Ra(CPURegister ra) {
     DCHECK_NE(ra.code(), kSPRegInternalCode);
@@ -2147,15 +2143,8 @@ class V8_EXPORT_PRIVATE Assembler : public AssemblerBase {
 
   // These encoding functions allow the stack pointer to be encoded, and
   // disallow the zero register.
-  static Instr RdSP(Register rd) {
-    DCHECK(!rd.IsZero());
-    return (rd.code() & kRegCodeMask) << Rd_offset;
-  }
-
-  static Instr RnSP(Register rn) {
-    DCHECK(!rn.IsZero());
-    return (rn.code() & kRegCodeMask) << Rn_offset;
-  }
+  static Instr RdSP(Register rd);
+  static Instr RnSP(Register rn);
 
   // Flags encoding.
   inline static Instr Flags(FlagsUpdate S);
